src/org/opensolaris/opengrok/util/OptionParser.java

```diff
@@ -0,0 +1,843 @@
+                }
+                optind = ++ii;
+            }
+        }
+        
+        // Prepare to gather any remaining arguments
+        // to send back to calling program.
+        
+        String[] remainingArgs = null;
+        
+        if (optind == -1) {
+            remainingArgs = args;
+        } else if (optind < args.length) {
+            remainingArgs = Arrays.copyOfRange(args, optind, args.length);
+        } else {
+            remainingArgs = new String[0];  // all args used up, send back empty.
+        }
+        
+        return remainingArgs;
+    }
+    
+    private String program() {
+        // Locate main program name from stack trace.
+        // The last element in the stack is where the 'main' entry point
+        // is located. The simple class name of this entry will be
+        // returned as the program name.
+        StackTraceElement[] stack = Thread.currentThread ().getStackTrace ();
+        StackTraceElement main = stack[stack.length - 1];
+        String[] name = main.getClassName().split("\\.");
+        return name[name.length-1];
+    }
+    
+    private String getPrologue() {
+        // Assign default prologue statement when none given.
+        if (prologue == null) {
+            prologue = "Usage: " + program() + " [options]";
+        }
+        
+        return prologue;
+    }
+    
+        
+    /**
+     * Define the prologue to be presented before the options summary.
+     * Example: Usage programName [options]
+     * @param text that makes up the prologue.
+     */
+    public void setPrologue(String text) {
+        prologue = text;
+    }
+    
+    /**
+     * Define the epilogue to be presented after the options summary.
+     * @param text that makes up the epilogue.
+     */
+    public void setEpilogue(String text) {
+        epilogue = text;
+    }
+    
+    /**
+     * Place text in option summary
+     * @param text to be inserted into option summary.
+     * 
+     * Example usage:
+     *  OptionParser opts = OptionParser.Do( parser -&gt; {
+     *   
+     *    parser.prologue = String.format("Usage: %s [options] bubba smith", program);
+     *    parser.separator("");
+     * 
+     *    parser.on("-y value", "--why me", "This is a description").Do( v -&gt; {
+     *        System.out.println("got " + v);
+     *    });
+     *       
+     *    parser.separator("  ----------------------------------------------");
+     *    parser.separator("  Common Options:");
+     *    ...
+     * 
+     *    parser.separator("  ----------------------------------------------");
+     *    parser.epilogue = "  That's all Folks!";
+     */
+    public void separator(String text) {
+        usageSummary.add(text);
+    }
+    
+    /**
+     * Obtain option summary
+     * @param indent a string to be used as the option summary initial indent.
+     * @return 
+     */
+    public String getUsage(String indent) {
```

Comment:
it would be really nice if this function could wrap the lines to terminal width (80 columns by default), it's one of the things that I find irritating with the current option parser.

Reply:
I wasn't expecting to get into reformatting here. That is one reason why the descriptions have multiple entries. The intent is that the programmer can format the option information to look exactly as it would on the command line in the program.  For example,

![image](https://user-images.githubusercontent.com/13596544/31670816-78581f16-b326-11e7-972e-c08df78355f8.png)

produces:

![image](https://user-images.githubusercontent.com/13596544/31670728-40055354-b326-11e7-871d-ffdbca2e5ab5.png)

You would not be able to guide the formatting as well (and as simply) if we start playing around moving text around programmatically.

Reply:
I agree this is a bit orthogonal to this change so file a new Issue for that pls.

Reply:
OK, I added issue #1824